Introduce nRF54H20 SoC and board

anangl opened the pull request:
Add initial support for the nRF54H20 SoC and definition of the nRF54H20 PDK board with targets that allow building for the Application, Radio, and Peripheral Processor (PPR) cores in this SoC.

57300 committed:
dts: Move nrf_common.dtsi to the common directory

... so that it can be included by ARM and RISC-V cores. For the same
reason, SysTick can no longer be disabled in this common file.

Signed-off-by: Grzegorz Swiderski <grzegorz.swiderski@nordicsemi.no>
Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

anangl committed:
drivers: serial: nrfx: Allow new UARTE instances to be used

Extend Kconfig definitions and nrfx_config translations so that UARTE
instances that are available in nRF54H20 can be used.

Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

anangl committed:
soc: nordic: Add initial support for nRF54H20 EngA

Add nrfx and Kconfig related infrastructure plus SoC initialization
code to allow building for nRF54H20 targets (Application and Radio
cores).

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>
Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

anangl committed:
boards: arm: Add initial support for nRF54H20 PDK

Add targets that allows building for the Application and Radio cores
in the nRF54H20 SoC on the nRF54H20 PDK board.

Signed-off-by: Grzegorz Swiderski <grzegorz.swiderski@nordicsemi.no>
Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>
Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

gmarull committed:
dt-bindings: misc: add nordic,nrf-ficr helper definitions

Add definitions to access all FICR fields in the nRF54H20 EngA SoC.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
scripts: west: runners: nrfjprog: add option to not erase

On some new SoCs, the erase option is not required, so introduce a new
option that doesn't set any erase mode.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
scripts: west: runners: nrfjprog: add erasepage op

Add a new operation that allows erasing pages.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
scripts: west: runners: nrfjprog: add basic support for nRF54H series

Add basic support to flash application and/or radio core for nRF54H
series. Note that features like merged hexes present in nRF53 series is
not supported.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

hubertmis commented on boards/arm/nrf54h20pdk_nrf54h20/nrf54h20pdk_nrf54h20_cpuapp.dts:
```diff
@@ -0,0 +1,132 @@
+	status = "okay";
+};
+
+&cpuapp_dma_region {
+	status = "okay";
+};
```
Does nrf-regtool understand `status = "okay"` and `status = "disabled"`? Or should we modify compatibles here instead?

hubertmis commented on boards/arm/nrf54h20pdk_nrf54h20/nrf54h20pdk_nrf54h20-memory_map.dtsi:
```diff
@@ -0,0 +1,79 @@
+ */
+
+/ {
+	reserved-memory {
```
Are ram0x and ram2x intentionally skipped because unused in the scope of this PR?

hubertmis commented on boards/arm/nrf54h20pdk_nrf54h20/nrf54h20pdk_nrf54h20-memory_map.dtsi:
```diff
@@ -0,0 +1,79 @@
+	};
+};
+
+&mram1x {
```
Are rw partitions and partitions reserved by the system intentionally skipped?

hubertmis commented on boards/arm/nrf54h20pdk_nrf54h20/nrf54h20pdk_nrf54h20-memory_map.dtsi:
```diff
@@ -0,0 +1,79 @@
+		};
+	};
+
+	cpurad_rx_partitions: cpurad-rx-partitions {
```
How about swapping `cpuapp_rx_partitions` and `cpurad_rx_partitions` order to keep partition addresses incrementing?

anangl replied:
Applied.

hubertmis commented on boards/arm/nrf54h20pdk_nrf54h20/nrf54h20pdk_nrf54h20_cpurad.dts:
```diff
@@ -0,0 +1,58 @@
+	owned-channels = <7 8 9 10 11 12 13 14>;
+};
+
+&uart135 {
+	pinctrl-0 = <&uart135_default>;
+	pinctrl-1 = <&uart135_sleep>;
+	pinctrl-names = "default", "sleep";
+};
```
Why this node is defined here if its status is "disabled"?

hubertmis replied:
Ok, now I'm confused. I see for the nrf54h20pdk_nrf54h20 both cpuapp and cpurad set `status = "okay"` for `&uart136` and both set `pinctrl` for `&uart135` with `status = "disabled"`. I expected cpuapp would set `status = "okay"` and `pinctrl` properties for one of the UART instances, while cpurad would do the same for the other one.
How the resources access conflict is resolved now?

57300 replied:
Right. In a multi-core application, cpurad would have to disable `uart136` and enable `uart135` to avoid conflicts.

I'm not sure which default configuration would be preferred in Zephyr:

1. Both cores use the same UART, so that if you run a basic sample on either one, it will output over the same pins.
2. Both cores use different UARTs, to support multi-core samples without extra overlays.

anangl replied:
In nRF5340 DK, we use separate pins (hence, separate VCOMs) for UART consoles on the Application and Network cores. We should probably use a similar scheme here and consequently use separate UARTs for these two cores.

anangl replied:
Changed to use separate UARTs and pins on Application and Radio cores.

carlocaione commented:
Is this supposed to be a draft? Many commit messages are just `WIP`.

FrancescoSer commented on boards/arm/nrf54h20pdk_nrf54h20/Kconfig.board:
```diff
@@ -0,0 +1,10 @@
+# Copyright (c) 2024 Nordic Semiconductor ASA
```
@anangl Please let me know when RST files will be in the PR

anangl replied:
> @anangl Please let me know when RST files will be in the PR

@FrancescoSer RST file is now present in the PR.

57300 committed:
modules: hal_nordic: Integrate nrf-regtool

nrf-regtool is a Python utility from Nordic Semiconductor, which is used
for generating binary files with register values for given peripherals.
It sources the descriptions of peripheral registers from CMSIS-SVD files
(typically ones bundled with nRF MDK).

For some peripherals, such as UICR, nrf-regtool supports parsing values
from devicetree as well, based on the bindings already found in Zephyr.

Currently, this tool is not submitted as a script to Zephyr, but it can
be installed from PyPI.

Having nrf-regtool installed is recommended when working with nRF54H20.
Booting the Application or Radiocore CPU requires flashing not only its
firmware, but also its respective UICR instance. On this SoC, the UICR
is used to assign ownership of global hardware resources, including
memory and peripherals, to individual cores. The Zephyr build system can
call nrf-regtool to generate the UICR based on devicetree, to reflect
the boot-time hardware configuration required for a given application.
The generated `uicr.hex` is then merged with `zephyr.hex`, so that they
can be flashed together using west.

The build system integration takes the form of a CMake package, which
includes a version check and reusable components; over time, some of
these components can be reused by sysbuild. This package is located in
the `hal_nordic` module, because it depends on the `SOC_SVD_FILE` CMake
variable, which is defined there as well.

Signed-off-by: Grzegorz Swiderski <grzegorz.swiderski@nordicsemi.no>

57300 committed:
modules: hal_nordic: Set SOC_SVD_FILE

This CMake variable contains a path to a CMSIS-SVD file from nRF MDK,
which describes the selected Nordic SoC.

Initially, this will be used as an argument to nrf-regtool.

Signed-off-by: Grzegorz Swiderski <grzegorz.swiderski@nordicsemi.no>

anangl commented:
Rebased.

carlescufi commented:
> Is this supposed to be a draft? Many commit messages are just `WIP`.

This is no longer the case @carlocaione. Feel free to review, thanks!

anangl commented:
> > Is this supposed to be a draft? Many commit messages are just `WIP`.
> 
> This is no longer the case @carlocaione. Feel free to review, thanks!

I only got rid of the commits from #67446 so far. These two WIP commits are still present. They will be replaced with proper ones in a while.

gmarull commented on dts/arm/nordic/nrf52805.dtsi:
```diff
@@ -320,3 +320,8 @@
 	arm,num-irq-priority-bits = <3>;
 };
+
+&systick {
+	/* Use RTC for system clock, instead of SysTick. */
+	status = "disabled";
+};
```
systick change in `dts: Move nrf_common.h to the common directory`, should be split.

anangl replied:
This is just moved here from `nrf_common.dtsi`, as this is an ARM specific thing so it can no longer stay there. I added a related note in the commit message.

gmarull commented on dts/arm/nordic/nrf51822.dtsi:
```diff
@@ -24,6 +24,7 @@
 		ficr: ficr@10000000 {
 			compatible = "nordic,nrf-ficr";
 			reg = <0x10000000 0x1000>;
+			#nordic,ficr-cells = <1>;
```
note: I've opened https://github.com/zephyrproject-rtos/zephyr/pull/68331 so we can simplify this PR a bit and focus on 54H only changes.

gmarull commented on dts/arm/nordic/nrf54h20_eng0_cpuapp.dtsi:
```diff
@@ -0,0 +1,38 @@
+
+/ {
+	soc {
+		compatible = "simple-bus";
```
can simple-bus, ranges be moved to the global dt file?

57300 replied:
The idea with setting `simple-bus` and `ranges` here is to indicate that this CPU is connected to this bus, while other CPUs may or may not be. In the global dt file, they should appear as "indirect buses", to borrow a system DT term.

gmarull replied:
ok, got it

gmarull commented on dts/arm/nordic/nrf54h20_eng0_cpuapp.dtsi:
```diff
@@ -0,0 +1,38 @@
+};
+
+&cpuapp_ppb {
+	compatible = "simple-bus";
+	ranges;
```
same

gmarull commented on dts/bindings/arm/nordic,nrf-uicr-v2.yaml:
```diff
@@ -0,0 +1,26 @@
+# Copyright (c) 2024 Nordic Semiconductor ASA
+# SPDX-License-Identifier: Apache-2.0
+
+description: Nordic UICR v2 (User Information Configuration Registers)
```
I'd suggest splitting bindings into its own commits, since this one is large enough.

anangl replied:
Done.

gmarull commented on dts/common/nordic/nrf54h20_eng0.dtsi:
```diff
@@ -0,0 +1,309 @@
+			compatible = "arm,cortex-m33";
+			reg = <2>;
+			device_type = "cpu";
+			clock-frequency = <DT_FREQ_M(320)>;
+		};
+
+		cpurad: cpu@3 {
+			compatible = "arm,cortex-m33";
+			reg = <3>;
+			device_type = "cpu";
+			clock-frequency = <DT_FREQ_M(256)>;
```
shouldn't we use `clocks = <&hsfll>;` here? otherwise we duplicate frequency information.

gmarull commented on soc/arm/nordic_nrf/CMakeLists.txt:
```diff
@@ -1,9 +1,11 @@
 # SPDX-License-Identifier: Apache-2.0
 
+zephyr_library()
```
nit: to simplify this PR, I'd suggest opening a PR with e879629a70252987bacdbf95162e7587274adcd5

anangl replied:
Done: #68340.

gmarull commented on soc/arm/nordic_nrf/nrf54h/CMakeLists.txt:
```diff
@@ -0,0 +1,5 @@
+# SPDX-License-Identifier: Apache-2.0
+
+zephyr_library_sources(
+  soc.c
+  )
```
`zephyr_library_sources(soc.c)`

gmarull commented on soc/arm/nordic_nrf/nrf54h/Kconfig.defconfig.nrf54h20_cpuapp:
```diff
@@ -0,0 +1,14 @@
+if SOC_NRF54H20_ENG0_CPUAPP
+
+config SOC
+	default "nRF54H20_CPUAPP"
```
`default "nrf54h20_cpuapp"`

gmarull replied:
I'd use lowercase here, as we do eg for soc family/series

gmarull replied:
Also, soc name should be changed to include `enga`, eg `nrf54h20_enga_cpuapp`

gmarull commented on soc/arm/nordic_nrf/nrf54h/Kconfig.defconfig.nrf54h20_cpurad:
```diff
@@ -0,0 +1,14 @@
+if SOC_NRF54H20_ENG0_CPURAD
+
+config SOC
+	default "nRF54H20_CPURAD"
```
`default "nrf54h20_cpurad"`

gmarull commented on soc/arm/nordic_nrf/nrf54h/Kconfig.soc:
```diff
@@ -0,0 +1,43 @@
+# SPDX-License-Identifier: Apache-2.0
+
+config SOC_NRF54H20
+	bool "NRF54H20"
```
`bool "nRF54H20"`

gmarull commented on soc/arm/nordic_nrf/nrf54h/soc.c:
```diff
@@ -0,0 +1,106 @@
+#define HSFLL_NODE DT_NODELABEL(cpuapp_hsfll)
+#elif defined(NRF_RADIOCORE)
+#define HSFLL_NODE DT_NODELABEL(cpurad_hsfll)
+#else
+#error "Unsupported SoC"
+#endif
```
this is likely redundant?

anangl replied:
Removed.

gmarull commented on soc/arm/nordic_nrf/nrf54h/soc.c:
```diff
@@ -0,0 +1,106 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/cache.h>
+#include <zephyr/devicetree.h>
+#include <zephyr/init.h>
+#include <zephyr/kernel.h>
+#include <zephyr/logging/log.h>
+#include <hal/nrf_hsfll.h>
+#include <hal/nrf_lrcconf.h>
+#include <soc/nrfx_coredep.h>
+LOG_MODULE_REGISTER(soc, CONFIG_SOC_LOG_LEVEL);
```
```c
#include <zephyr/cache.h>
#include <zephyr/devicetree.h>
#include <zephyr/init.h>
#include <zephyr/kernel.h>
#include <zephyr/logging/log.h>

#include <hal/nrf_hsfll.h>
#include <hal/nrf_lrcconf.h>
#include <soc/nrfx_coredep.h>

LOG_MODULE_REGISTER(soc, CONFIG_SOC_LOG_LEVEL);
```

gmarull commented on soc/arm/nordic_nrf/nrf54h/soc.c:
```diff
@@ -0,0 +1,106 @@
+
+static void power_domain_init(void)
+{
+#if defined(NRF_LRCCONF010)
+	/*
+	 * Set:
+	 *  - LRCCONF010.POWERON.MAIN: 1
+	 *  - LRCCONF010.POWERON.ACT: 1
+	 *  - LRCCONF010.RETAIN.MAIN: 1
+	 *  - LRCCONF010.RETAIN.ACT: 1
+	 *
+	 *  This is done here at boot so that when the idle routine will hit
+	 *  WFI the power domain will be correctly retained.
+	 */
+
+	nrf_lrcconf_poweron_force_set(NRF_LRCCONF010, NRF_LRCCONF_POWER_MAIN, true);
+	nrf_lrcconf_poweron_force_set(NRF_LRCCONF010, NRF_LRCCONF_POWER_DOMAIN_0, true);
+
+	nrf_lrcconf_retain_set(NRF_LRCCONF010, NRF_LRCCONF_POWER_MAIN, true);
+	nrf_lrcconf_retain_set(NRF_LRCCONF010, NRF_LRCCONF_POWER_DOMAIN_0, true);
+#endif
+
+#if defined(NRF_LRCCONF000) && defined(CONFIG_SOC_NRF54H20_ENG0_CPUAPP)
```
question: are NRF_LRCCONF010/NRF_LRCCONF000 always defined?

anangl replied:
It seems they are. I'll drop these checks.

gmarull commented on soc/arm/nordic_nrf/nrf54h/soc.h:
```diff
@@ -0,0 +1,17 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @file SoC configuration macros for the
+ *       Nordic Semiconductor nRF54H family processors.
+ */
```
this can be dropped, this file is not taken by Doxygen

anangl replied:
Applied.

gmarull commented on tests/lib/cpp/cxx/testcase.yaml:
```diff
@@ -34,8 +34,10 @@ tests:
     arch_exclude: posix
-    # Exclude nRF54L15 as its HAL is not compatible with C++98.
-    platform_exclude: nrf54l15pdk_nrf54l15_cpuapp
+    # Exclude nRF54L15 and nRF54H20 as Nordic HAL is not compatible with C++98.
+    platform_exclude:
+      - nrf54l15pdk_nrf54l15_cpuapp
+      - nrf54h20pdk_nrf54h20_cpuapp
```
this probably should go into its own commit

anangl replied:
Done.

gmarull commented on dts/common/nordic/nrf_common.dtsi:
```diff
@@ -8,6 +8,7 @@
 #include <zephyr/dt-bindings/adc/nrf-adc.h>
 #include <zephyr/dt-bindings/gpio/gpio.h>
 #include <zephyr/dt-bindings/i2c/i2c.h>
+#include <zephyr/dt-bindings/input/input-event-codes.h>
```
nit: maybe split, it feels weird to see this in board commit (even though it's required)

anangl replied:
Done.

gmarull commented on modules/hal_nordic/CMakeLists.txt:
```diff
@@ -6,3 +6,19 @@ if (CONFIG_NRF_802154_RADIO_DRIVER OR CONFIG_NRF_802154_SERIALIZATION)
+  list(APPEND nrf_regtool_components GENERATE:UICR)
+endif()
+if(DEFINED nrf_regtool_components)
+  find_package(nrf-regtool 5.0.1
+    COMPONENTS ${nrf_regtool_components}
+    PATHS ${CMAKE_CURRENT_LIST_DIR}/nrf-regtool
+    NO_CMAKE_PATH
+    NO_CMAKE_ENVIRONMENT_PATH
+    NO_SYSTEM_ENVIRONMENT_PATH
+    NO_CMAKE_PACKAGE_REGISTRY
+    NO_CMAKE_SYSTEM_PATH
+    NO_CMAKE_SYSTEM_PACKAGE_REGISTRY
+  )
+endif()
```
this now issues a warning if not found, however, the final hex file is in practice unusable if regtool doesn't populate a UICR file. I think we should make this `REQUIRED`.

57300 replied:
I'd say it's not strictly required, because the user is free to provide their own UICR file separately. You can also turn off `CONFIG_NRF_REGTOOL_GENERATE_UICR`, which would hide the warning.

gmarull replied:
> I'd say it's not strictly required, because the user is free to provide and flash their own UICR file separately. You can also turn off `CONFIG_NRF_REGTOOL_GENERATE_UICR`, which would hide the warning.

but is this a realistic case? My fear is that a warning is easy to ignore, and then, flashing the app will just not work.

57300 replied:
It's not very realistic, so we could say we don't support it, and make it `REQUIRED`.

anangl committed:
tests: arm_irq_vector_table: Add special handling for nRF54H20

Different set of IRQ lines need to be used for this SoC and the CLOCK
IRQ is not to be installed in the vector table.

Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

fkokosinski commented on dts/common/nordic/nrf54h20_eng0.dtsi:
```diff
@@ -0,0 +1,309 @@
+		};
+
+		cpuppr: cpu@d {
+			compatible = "nordic,riscv";
```
Let's add a separate binding in `dts/bindings/cpu`  that inherits from the common `riscv,cpus.yaml`, so that the `riscv,isa` property is properly enforced

anangl replied:
Applied.

gmarull committed:
drivers: misc: nordic_vpr_launcher: initial version

Add a custom driver that takes care of loading and launching RISC-V VPR
cores found on the new nRF54 SoCs.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

anangl committed:
drivers: serial: Kconfig.nrfx: Filter out options unsupported on nRF54H20

On nRF54H20, only the new shim can be used and the enhanced poll out
cannot be enabled since there is no DPPI support for this SoC yet.

Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

anangl committed:
drivers: timer: nrf_grtc_timer: Add dependency on nRF clock control

... in the related parts, so that the driver can be used on nRF54H20
where the clock control is not present yet.

Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

anangl committed:
tests: lib: cpp: Exclude cpp98 test on nRF54H20

Similar to nRF54L15, the MDK files in HAL for this SoC are not
compatible with C++98, so currently the test cannot be performed
on this SoC.

Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

anangl commented:
Rebased to remove commits that were merged in #68331 and #68340.

anangl committed:
dts: Add and extend Nordic bindings needed for nRF54H20

Add a set of bindings that will be used in the nRF54H20 SoC definition.
Extend the existing GPIOTE binding with properties needed for this SoC.
Also do a tiny clean-up in the bindings added recently for nRF54L15
(HFXO and LFXO).

Signed-off-by: Grzegorz Swiderski <grzegorz.swiderski@nordicsemi.no>
Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>
Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

anangl committed:
dts: Add initial support for nRF54H20 EngA SoC

Add definition of the nRF54H20 SoC revision EngA with its Application,
Radio, and Peripheral Processor (PPR) cores and basic peripherals:
GRTC, GPIOs, GPIOTE, and UARTs.

Signed-off-by: Grzegorz Swiderski <grzegorz.swiderski@nordicsemi.no>
Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>
Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

anangl committed:
modules: hal_nordic: nrfx_glue: Include cmsis_core_m_defaults.h

... to cover missing __ICACHE_PRESENT and __DCACHE_PRESENT symbols that
should be defined in MDK files.

Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

gmarull committed:
soc: riscv: nordic_nrf: add initial support for VPR core

Add initial support for the VPR RISC-V core found in the new nRF54 SoCs.

Signed-off-by: Carlo Caione <ccaione@baylibre.com>
Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
boards: riscv: add nrf54h20pdk_nrf54h20_cpuppr

Add a board that allows to build for the nRF54H20 PPR RISC-V core.

Signed-off-by: Grzegorz Swiderski <grzegorz.swiderski@nordicsemi.no>
Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
modules: hal_nordic: add configuration for nRF54H PPR core

Add nrfx configuration file to make the nrfx HAL work for
nRF54H PPR core.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>
Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

gmarull committed:
drivers: serial: nrfx_uarte2: drop soc.h

As it is not required (e.g. RISC-V nRF54H port does not provide soc.h)

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

anangl committed:
dts: nordic: Include input-event-codes.h from nrf_common.dtsi

... so that there is no need to include that header individually
for every added board based on an nRF SoC.

Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

anangl committed:
modules: hal_nordic: Use common nrfx_config section for GRTC

Move code that prepares `NRFX_CONFIG_GRTC_*` definitions based on
information from devicetree from the nRF54L15 nrfx_config header
to the global one, so that the code can be used by nRF54H20, too.

The checks that validate owned-channels and child-owned-channels
DT properties are moved to the nrf_grtc_timer driver so that
the global nrfx_config is not polluted unnecessarily.

The default values in nrfx_config_nrf54l15_enga_application.h
are restored to those from the corresponding template file.

Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

gmarull committed:
scripts: kconfig: functions: add dt_chosen_partition_addr_int|hex

dt_chosen_partition_addr_int|hex allow obtaining the absolute address of
a partition, which is the result of the grandparent node address plus
the partition node address.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
soc: riscv: nordic_nrf: nrf54h: introduce PPR support

Add support for the nRF54H PPR (Peripheral Processor), based on the VPR
RISC-V core.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
soc: common: nordic_nrf: move pinctrl_soc.h to a common dir

Because both, RISC-V and ARM cores share the same pinctrl driver. The
top level common folder will disappear with the introduction of HWMv2,
where multi-arch SoCs will be well supported.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull committed:
soc: arm: nordic_nrf: move Kconfig.peripherals to common folder

Because RISC-V cores also need to include this file, so it is no longer
ARM specific.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull commented on dts/bindings/cpu/nordic,riscv.yaml:
```diff
@@ -0,0 +1,8 @@
+
+description: Nordic Semiconductor RISC-V CPU
+
+compatible: "nordic,riscv"
```
nit: should be `nordic,vpr`

57300 replied:
Or `nordic,nrf-vpr`?

gmarull replied:
> Or `nordic,nrf-vpr`?

yup

gmarull replied:
I'd actually stick to `nordic,vpr`, since it's a RISC-V core that could be used out of nRF products, who knows.

gmarull commented on dts/bindings/cpu/nordic,riscv.yaml:
```diff
@@ -0,0 +1,8 @@
+# Copyright (c) 2024 Nordic Semiconductor ASA
+# SPDX-License-Identifier: Apache-2.0
+
+description: Nordic Semiconductor RISC-V CPU
```
`Nordic Semiconductor RISC-V VPR CPU`

gmarull commented on dts/bindings/riscv/nordic,vpr-coprocessor.yaml:
```diff
@@ -0,0 +1,32 @@
+    description: |
+      Processor ID of the VPR core.
+
+  loader-img-dst:
+    type: phandle
+    required: true
+    description: |
+      Destination memory handle. Points to the area from which the VPR core
+      will start execution.
+
+  loader-img-src:
+    type: phandle
+    description: |
+      Source memory handle. If specified, then the VPR's firmware image will be
+      relocated from this location to the one given by "loader-img-dst".
```
please apply https://github.com/teslabs/zephyr/commit/809e8fb7a23ef572d81ee1f19ef50a5f32e38152

gmarull commented on boards/arm/nrf54h20pdk_nrf54h20/nrf54h20pdk_nrf54h20_cpuapp.dts:
```diff
@@ -0,0 +1,132 @@
+	status = "okay";
+};
+
+&cpuapp_rx_partitions {
```
in board dts file, this can be added https://github.com/teslabs/zephyr/commit/413253cdd77e6e51342f822e02682b96ad8f81eb

gmarull commented on dts/common/nordic/nrf54h20_enga.dtsi:
```diff
@@ -0,0 +1,309 @@
+/*
```
please add https://github.com/teslabs/zephyr/commit/4844be1a8653a34cc407eeca831224dd37fdcea8

gmarull replied:
please don't :-)

masz-nordic commented on boards/arm/nrf54h20pdk_nrf54h20/nrf54h20pdk_nrf54h20_cpuapp_defconfig:
```diff
@@ -0,0 +1,32 @@
+# Enable hardware stack protection
+CONFIG_HW_STACK_PROTECTION=y
+
+# Enable cache
```
```suggestion
# Enable instruction cache
```

masz-nordic replied:
Same in Radio core.

anangl replied:
I moved disabling of DCACHE to `Kconfig.defconfig`, so that it can be overridden if needed. Please, take a look.

masz-nordic commented on dts/bindings/interrupt-controller/nordic,nrf-clic.yaml:
```diff
@@ -0,0 +1,19 @@
+# Copyright (c) 2024 Nordic Semiconductor ASA
+# SPDX-License-Identifier: Apache-2.0
+
+description: Nordic CLIC
```
```suggestion
description: Nordic VPR CLIC
```

masz-nordic commented on dts/bindings/riscv/nordic,vpr-coprocessor.yaml:
```diff
@@ -0,0 +1,32 @@
+description: |
+  VPR coprocessor
+
+  VPR is a RISC-V CPU implementation. VPR instances are exposed to other CPUs as
+  peripherals. A domain can take ownership of one or more VPRs.
```
```suggestion
  VPR is a RISC-V CPU implementation.
```
Not true for all VPRs.

gmarull replied:
We should probably name this compatible `nordic,nrf-vpr-coprocessor`, as it's for the VPR seen as peripheral from nRF SoCs (not the core itself, which has its own binding).

masz-nordic replied:
My understanding is that this compatible is used to describe two things at the same time:
- how to configure security and ownership (nRF54H specific)
- how to run VPR (applies to both nRF54H and nRF54L)

Perhaps the proper way would be to split those into separate bindings?

57300 replied:
> * how to configure security and ownership (nRF54H specific)

Not really. nrf-regtool doesn't need an extra binding for that specifically, when it comes to VPRs. The only thing specific to nRF54H is the `cpu` property, which is really just hardware description.

masz-nordic replied:
Ok, the issue with H-specific stuff is addressed.
But I think the description of node should be changed - at least last sentence has to be removed.

anangl replied:
Removed the last sentence.

masz-nordic commented on dts/bindings/riscv/nordic,vpr-coprocessor.yaml:
```diff
@@ -0,0 +1,32 @@
+include: base.yaml
+
+properties:
+  cpu:
+    type: int
+    required: true
+    description: |
+      Processor ID of the VPR core.
```
Is it used anywhere?

gmarull replied:
IIRC by regtool @57300 ?

masz-nordic replied:
Then it should be optional, it's not needed in nRF54L.

gmarull replied:
we should probably add a link to the CPU node instead, this duplicates unnecessary information, but I assume this requires fixes to regtool...

57300 replied:
Can't link the CPU node, because it has to be deleted.

57300 replied:
Let's make it optional. I wasn't aware that nRF54L doesn't define Processor IDs.

anangl replied:
Done.

gmarull committed:
snippets: add nordic-ppr

Add a new snippet that allows to build any application with the
capability to boot the PPR core found in some nRF54 SoCs.

Signed-off-by: Gerard Marull-Paretas <gerard@teslabs.com>

gmarull commented on include/zephyr/dt-bindings/misc/nordic-nrf-ficr-nrf54h20-enga.h:
```diff
@@ -0,0 +1,96 @@
+/*
```
nit: adjust commit message for edf118179c7ac31a975dff996a8d77f85802f7b6 (now all FICR fields are there, not only HSFLL)

gmarull commented on dts/common/nordic/nrf54h20_enga.dtsi:
```diff
@@ -0,0 +1,310 @@
+		};
+	};
+
+	clocks {
+		fll16m: fll16m {
+			compatible = "fixed-clock";
+			#clock-cells = <0>;
+			clock-frequency = <DT_FREQ_M(16)>;
+		};
+	};
```
question for @57300, does PPR see FLL16M? If not maybe this node should be deleted in ppr dtsi

57300 replied:
> does PPR see FLL16M?

It's in the Global domain, so I think it's fine to leave it.

gmarull commented on soc/arm/nordic_nrf/nrf54h/soc.h:
```diff
@@ -0,0 +1,12 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef _NORDICSEMI_NRF54H_SOC_H_
```
`#ifndef SOC_ARM_NORDIC_NRF_NRF54H_SOC_H_`

gmarull commented on modules/hal_nordic/nrfx/CMakeLists.txt:
```diff
@@ -159,3 +159,25 @@ zephyr_compile_definitions_ifdef(CONFIG_SOC_NRF54LX_SKIP_CLOCK_CONFIG NRF_SKIP_C
+mdk_svd_ifdef(CONFIG_SOC_NRF5340_CPUAPP       nrf5340_application.svd)
+mdk_svd_ifdef(CONFIG_SOC_NRF5340_CPUNET       nrf5340_network.svd)
+mdk_svd_ifdef(CONFIG_SOC_NRF54H20_ENGA_CPUAPP nrf54h20_enga_application.svd)
+mdk_svd_ifdef(CONFIG_SOC_NRF54H20_ENGA_CPURAD nrf54h20_enga_radiocore.svd)
```
I forgot to add cpuppr in the ppr commit

gmarull commented on scripts/west_commands/runners/nrf_common.py:
```diff
@@ -24,11 +24,26 @@
 ErrNotAvailableBecauseProtection = 24
```
nit: drop `Flashing other cores like PPR/VPR is not supported yet` from f4be9e7ec2c05a2520547fcce65b2d70177fc004 message as it just works

gmarull commented on boards/arm/nrf54h20pdk_nrf54h20/nrf54h20pdk_nrf54h20_cpuapp.dts:
```diff
@@ -97,6 +97,11 @@
 	status = "okay";
 };
 
+&cpuppr_vpr {
```
660e128a38b6ee367ed1755e0fbdb7eebcbc5c79 may be merged with the board commit

gmarull commented on snippets/nordic-ppr/README.rst:
```diff
@@ -0,0 +1,10 @@
+.. _nordic-ppr:
```
nit: commit message typo `snipper` -> `snippet`

gmarull commented on modules/hal_nordic/nrfx/nrfx_config.h:
```diff
@@ -748,6 +748,8 @@
     #include <nrfx_config_nrf54h20_enga_application.h>
```
nit: drop `and CMake injected definitions` 67c5c07db0962c08ceb412c7715888d3925464c0 as it is done later

gmarull commented on boards/riscv/nrf54h20pdk_nrf54h20/Kconfig.board:
```diff
@@ -0,0 +1,6 @@
+# Copyright (c) 2024 Nordic Semiconductor ASA
```
nit 3ca4a42f7b9aa99473114e3dd9d42b3b57a6a308 should be signed off by @57300  as well

gmarull commented on soc/riscv/nordic_nrf/CMakeLists.txt:
```diff
@@ -0,0 +1,4 @@
+# Copyright (c) 2024 Nordic Semiconductor
```
@carlocaione may want a signed off for 2d93018da19ae962087097ed08ed9a31994d42d9 :-)

gmarull commented on dts/bindings/riscv/nordic,vpr-coprocessor.yaml:
```diff
@@ -0,0 +1,29 @@
+# Copyright (c) 2024 Nordic Semiconductor ASA
+# SPDX-License-Identifier: Apache-2.0
+
+compatible: "nordic,vpr-coprocessor"
```
if adjusted, vpr loader driver needs to be adjusted as well

masz-nordic commented on dts/common/nordic/nrf54h20_enga.dtsi:
```diff
@@ -0,0 +1,310 @@
+				#size-cells = <1>;
+				ranges = <0x0 0x908000 0x4000>;
+
+				cpuppr_clic: interrupt-controller@1000 {
+					compatible = "nordic,nrf-clic";
+					reg = <0x1000 0x3000>;
+					status = "disabled";
+					#interrupt-cells = <2>;
+					interrupt-controller;
+					#address-cells = <1>;
+				};
```
Not sure if this is the right place for this node.
Maybe `soc` in `dts/riscv/nordic/nrf54h20_enga_cpuppr.dtsi` would be better?

anangl replied:
@57300 ^

masz-nordic replied:
I'm not sure what would be the benefit in routing it in owner's DT instead of VPR's DT - but I might be missing something specific to nRF54H architecture.
Not a blocker from my side anyway - we will revisit it when adding VPR to nRF54L.

masz-nordic commented on modules/hal_nordic/nrfx/nrfx_config_nrf54l15_enga_application.h:
```diff
@@ -249,37 +249,21 @@
-#endif /* DT_HAS_COMPAT_STATUS_OKAY(nordic_nrf_grtc) */
-
-#endif /* NRFX_GRTC_CONFIG_ALLOWED_CC_CHANNELS_MASK */
+#define NRFX_GRTC_CONFIG_ALLOWED_CC_CHANNELS_MASK 0x000007ff
```
Why is it hardcoded now?

anangl replied:
It is not hardcoded. The mask generation was moved one level up, to `nrfx_config.h`. The above value will be used only when the GRTC node is disabled in devicetree, so when GRTC is used directly with the nrfx_grtc driver, but then this mask should be set accordingly by the application that uses GRTC this way.

FrancescoSer commented on boards/arm/nrf54h20pdk_nrf54h20/doc/index.rst:
```diff
@@ -0,0 +1,148 @@
+
+nRF54H20 SoC provides support for the following devices:
+
+* :abbr:`ADC (Analog to Digital Converter)`
+* CLOCK
+* :abbr:`GPIO (General Purpose Input Output)`
+* :abbr:`GPIOTE (General Purpose Input Output tasks and events)`
+* :abbr:`GRTC (Global real-time counter)`
+* :abbr:`I2C (Inter-Integrated Circuit)`
+* MRAM
+* :abbr:`PWM (Pulse Width Modulation)`
+* RADIO (Bluetooth Low Energy and 802.15.4)
+* :abbr:`SPI (Serial Peripheral Interface)`
+* :abbr:`UART (Universal asynchronous receiver-transmitter)`
+* :abbr:`USB (Universal Serial Bus)`
+* :abbr:`WDT (Watchdog Timer)`
```
I guess this list should contain only : GPIO, GPIOTE, GRTC, UART (as in the table starting from line 60)

anangl replied:
This is what is available in the SoC, in hardware. That tables below list what is currently supported in Zephyr.

fabiobaltieri reviewed (approved):
Cool, had a scan through the PRs, not much point nitpicking, everything seems to me very well isolated to the nRF platforms and seems like a solid base, good stuff.

FrancescoSer commented on boards/arm/nrf54h20pdk_nrf54h20/doc/index.rst:
```diff
@@ -0,0 +1,148 @@
+The nRF54H20 PDK is a single-board preview development kit for evaluation
+and development on the Nordic nRF54H20 System-on-Chip (SoC).
+
+The nRF54H20 is a multicore SoC with:
```
```suggestion
The nRF54H20 is a multicore SoC that includes the following cores:
```

FrancescoSer reviewed (commented):
Just left a pair of suggestions

pdunaj commented on soc/arm/nordic_nrf/nrf54h/soc.c:
```diff
@@ -0,0 +1,104 @@
+static int nordicsemi_nrf54h_init(void)
+{
+#if defined(CONFIG_NRF_ENABLE_ICACHE)
+	sys_cache_instr_enable();
```
we don't add data cache enable at this point?
